mono/metadata/icall.c

```diff
@@ -1719,23 +1720,28 @@ type_from_parsed_name (MonoTypeNameParse *info, MonoStackCrawlMark *stack_mark,
 	} else {
 		assembly = mono_runtime_get_caller_from_stack_mark (stack_mark);
 	}
-	if (assembly) {
-		type_resolve = TRUE;
-		rootimage = assembly->image;
-	} else {
-		g_warning (G_STRLOC);
-	}
-	*caller_assembly = assembly;
 
-	if (info->assembly.name)
-		assembly = mono_assembly_load (&info->assembly, assembly ? assembly->basedir : NULL, NULL);
+	g_assert (assembly);
 
-	if (assembly) {
-		/* When loading from the current assembly, AppDomain.TypeResolve will not be called yet */
-		type = mono_reflection_get_type_checked (rootimage, assembly->image, info, ignoreCase, TRUE, &type_resolve, error);
-		goto_if_nok (error, fail);
+	type_resolve = TRUE;
+	rootimage = assembly->image;
+
+	g_assert (rootimage);
+
+	*caller_assembly = assembly;
+
+	if (info->assembly.name) {
+		MonoAssemblyByNameRequest req;
+		mono_assembly_request_prepare_byname (&req, MONO_ASMCTX_DEFAULT, alc);
+		req.requesting_assembly = assembly;
+		req.basedir = assembly ? assembly->basedir : NULL;
+		assembly = mono_assembly_request_byname (&info->assembly, &req, NULL);
```

Comment:
This can fail, right?  So then when you call `mono_reflection_get_type_checked (..., assembly->image, ....)` I think you could be dereferencing a null pointer.

I think you need to keep the `if (assembly) { ... }` stuff from before.